// docker-registry-downloader/lib/src/docker-registry/index.js
"use strict";
var $___46__46__47_fs_45_helper__;
var FsHelper = ($___46__46__47_fs_45_helper__ = require("../fs-helper"), $___46__46__47_fs_45_helper__ && $___46__46__47_fs_45_helper__.__esModule && $___46__46__47_fs_45_helper__ || {default: $___46__46__47_fs_45_helper__}).default;
var request = require('requestretry');
var Q = require('q');
var log = require('../helpers/logger');
var fs = require('fs');
var path = require('path');
var prettyBytes = require('pretty-bytes');
var _ = require('lodash');
var fsHelper = new FsHelper();
var DockerRegistry = function DockerRegistry(request_options) {
  this.__request_options = request_options || {};
};
($traceurRuntime.createClass)(DockerRegistry, {
  get request_options() {
    return this.__request_options;
  },
  set request_options(value) {
    this.__request_options = value;
  },
  tags: function(hubResult) {
    var request_options_local = this.__request_options;
    return new Q.Promise(function(resolve, reject) {
      var options = _.assign({
        url: 'https://' + hubResult.endpoint + '/v1/repositories/' + hubResult.namespace + '/' + hubResult.repository + '/' + 'tags',
        headers: {'Authorization': 'Token ' + hubResult.token}
      }, request_options_local);
      function callback(error, response, body) {
        if (!error && response.statusCode === 200) {
          var result = JSON.parse(body);
          log.debug('\n\n:: docker-registry - tags ::');
          log.debug(result);
          resolve(result);
        } else {
          reject(error);
        }
      }
      request(options, callback);
    });
  },
  getImageIdByTag: function(hubResult, tag) {
    var request_options_local = this.__request_options;
    return new Q.Promise(function(resolve, reject) {
      var options = _.assign({
        url: 'https://' + hubResult.endpoint + '/v1/repositories/' + hubResult.namespace + '/' + hubResult.repository + '/tags/' + tag,
        headers: {'Authorization': 'Token ' + hubResult.token}
      }, request_options_local);
      function callback(error, response, body) {
        if (!error && response.statusCode === 200) {
          var result = JSON.parse(body);
          log.debug('\n\n:: docker-registry - getImageIdByTag ::');
          log.debug(result);
          resolve(result);
        } else {
          reject(error);
        }
      }
      request(options, callback);
    });
  },
  ancestry: function(hubResult, imageId) {
    var request_options_local = this.__request_options;
    return new Q.Promise(function(resolve, reject) {
      var options = _.assign({
        url: 'https://' + hubResult.endpoint + '/v1/images/' + imageId + '/ancestry',
        headers: {'Authorization': 'Token ' + hubResult.token}
      }, request_options_local);
      function callback(error, response, body) {
        if (!error && response.statusCode === 200) {
          var result = JSON.parse(body);
          log.debug('\n\n:: docker-registry - ancestry ::');
          log.debug(result);
          resolve(result);
        } else {
          reject(error);
        }
      }
      try {
        request(options, callback);
      } catch (err) {
        reject(err);
      }
    });
  },
  imageJson: function(hubResult, imageId) {
    var request_options_local = this.__request_options;
    return new Q.Promise(function(resolve, reject) {
      var options = _.assign({
        url: 'https://' + hubResult.endpoint + '/v1/images/' + imageId + '/json',
        headers: {'Authorization': 'Token ' + hubResult.token}
      }, request_options_local);
      function callback(error, response, body) {
        if (!error && response.statusCode === 200) {
          var result = JSON.parse(body);
          log.debug('\n\n:: docker-registry - imageJson ::');
          log.debug(result);
          resolve(result);
        } else {
          reject(error);
        }
      }
      request(options, callback);
    });
  },
  allAnscestorByTag: function(hubResult, tag) {
    return Q.async($traceurRuntime.initGeneratorFunction(function $__2() {
      var imageId,
          anscestors;
      return $traceurRuntime.createGeneratorInstance(function($ctx) {
        while (true)
          switch ($ctx.state) {
            case 0:
              log.debug('\n\n:: docker-registry - allAnscestorByTag ::');
              log.debug('endpoint:', hubResult.endpoint);
              log.debug('token:', hubResult.token);
              log.debug('namespace:', hubResult.namespace);
              log.debug('repository:', hubResult.repository);
              log.debug('tag:', tag);
              log.debug('>>------------');
              $ctx.state = 12;
              break;
            case 12:
              $ctx.state = 2;
              return this.getImageIdByTag(hubResult, tag);
            case 2:
              imageId = $ctx.sent;
              $ctx.state = 4;
              break;
            case 4:
              $ctx.state = 6;
              return this.ancestry(hubResult, imageId);
            case 6:
              anscestors = $ctx.sent;
              $ctx.state = 8;
              break;
            case 8:
              $ctx.returnValue = anscestors;
              $ctx.state = -2;
              break;
            default:
              return $ctx.end();
          }
      }, $__2, this);
    }).bind(this))();
  },
  downloadImageGetSize: function(hubResult, imageId) {
    var request_options_local = this.__request_options;
    return function(callback) {
      var options = _.assign({
        url: 'https://' + hubResult.endpoint + '/v1/images/' + imageId + '/layer',
        headers: {'Authorization': 'Token ' + hubResult.token},
        method: 'GET'
      }, request_options_local);
      var r = request(options).on('response', function(res) {
        log.debug('\n\n:: docker-registry - downloadImageGetSize headers ::');
        log.debug('ID:  ', imageId);
        var len = parseInt(res.headers['content-length'], 10);
        log.debug('res.headers:', res.headers);
        log.debug('size:', prettyBytes(len));
        r.abort();
        callback(null, len);
      });
    };
  },
  downloadImage: function(hubResult, outputPath, imageId, iProgress) {
    var request_options_local = this.__request_options;
    return new Q.Promise(function(resolve) {
      var options = _.assign({
        url: 'https://' + hubResult.endpoint + '/v1/images/' + imageId + '/layer',
        headers: {'Authorization': 'Token ' + hubResult.token},
        method: 'GET'
      }, request_options_local);
      request(options).on('response', function(res) {
        log.debug('\n\n:: docker-registry - downloadImage headers ::');
        log.debug(res.headers);
        res.on('data', function(chunk) {
          if (iProgress) {
            iProgress(chunk.length);
          }
        });
        res.on('end', function() {
          resolve(outputPath);
        });
      }).pipe(fs.createWriteStream(outputPath));
    });
  },
  prepareLoading: function(hubResult, outputPath, imageId, iProgress) {
    return Q.async($traceurRuntime.initGeneratorFunction(function $__2() {
      var outputLoadPath,
          versionFilePath,
          jsonResult,
          jsonFilePath,
          layerTarFilePath;
      return $traceurRuntime.createGeneratorInstance(function($ctx) {
        while (true)
          switch ($ctx.state) {
            case 0:
              outputLoadPath = path.join(outputPath, imageId);
              $ctx.state = 32;
              break;
            case 32:
              $ctx.state = 2;
              return fsHelper.createCleanFolder(outputLoadPath);
            case 2:
              $ctx.maybeThrow();
              $ctx.state = 4;
              break;
            case 4:
              versionFilePath = path.join(outputLoadPath, "VERSION");
              $ctx.state = 34;
              break;
            case 34:
              $ctx.state = 6;
              return Q.nfcall(fs.writeFile, versionFilePath, "1.0");
            case 6:
              $ctx.maybeThrow();
              $ctx.state = 8;
              break;
            case 8:
              $ctx.state = 10;
              return this.imageJson(hubResult, imageId);
            case 10:
              jsonResult = $ctx.sent;
              $ctx.state = 12;
              break;
            case 12:
              jsonFilePath = path.join(outputLoadPath, "json");
              $ctx.state = 36;
              break;
            case 36:
              $ctx.state = 14;
              return Q.nfcall(fs.writeFile, jsonFilePath, JSON.stringify(jsonResult, ' ', 3));
            case 14:
              $ctx.maybeThrow();
              $ctx.state = 16;
              break;
            case 16:
              layerTarFilePath = path.join(outputLoadPath, "layer.tar");
              $ctx.state = 38;
              break;
            case 38:
              $ctx.state = 18;
              return this.downloadImage(hubResult, layerTarFilePath, imageId, iProgress);
            case 18:
              $ctx.maybeThrow();
              $ctx.state = 20;
              break;
            case 20:
              $ctx.state = 22;
              return fsHelper.tarPack(outputLoadPath, path.join(outputLoadPath, '..', imageId + '.tar'));
            case 22:
              $ctx.maybeThrow();
              $ctx.state = 24;
              break;
            case 24:
              $ctx.state = 26;
              return fsHelper.removeDirRecursive(outputLoadPath);
            case 26:
              $ctx.maybeThrow();
              $ctx.state = 28;
              break;
            case 28:
              $ctx.returnValue = outputLoadPath;
              $ctx.state = -2;
              break;
            default:
              return $ctx.end();
          }
      }, $__2, this);
    }).bind(this))();
  }
}, {});
module.exports = {
  __esModule: true,
  get default() {
    return DockerRegistry;
  }
};
