// docker-registry-downloader/lib/src/docker-remote/index.js
"use strict";
var Docker = require('dockerode');
var fs = require('fs');
var Q = require('q');
var log = require('../helpers/logger');
var path = require('path');
var _ = require('lodash');
var DockerRemote = function DockerRemote(dockerode_options) {
  if (dockerode_options && dockerode_options.dockerode_modem) {
    this.docker = new Docker();
    this.docker.modem = dockerode_options.dockerode_modem;
  } else {
    var socket = process.env.DOCKER_SOCKET || (dockerode_options && dockerode_options.socket_dockerode) || '/var/run/docker.sock';
    var stats = fs.statSync(socket);
    if (!stats.isSocket()) {
      throw new Error("Are you sure the docker is running?");
    }
    this.docker = new Docker({socketPath: socket});
  }
};
($traceurRuntime.createClass)(DockerRemote, {
  listActiveContainers: function() {
    return new Q.Promise(function(resolve, reject) {
      try {
        this.docker.listContainers({all: false}, function(err, containers) {
          log.debug('\n\n:: docker-remote - listActiveContainers ::');
          log.debug(containers);
          resolve(containers);
        });
      } catch (err) {
        log.error(err);
        reject(err);
      }
    }.bind(this));
  },
  listAllContainers: function() {
    return new Q.Promise(function(resolve, reject) {
      try {
        this.docker.listContainers({all: true}, function(err, containers) {
          log.debug('\n\n:: docker-remote - listAllContainers ::');
          log.debug(containers);
          resolve(containers);
        });
      } catch (err) {
        log.error(err);
        reject(err);
      }
    }.bind(this));
  },
  listImages: function() {
    return new Q.Promise(function(resolve, reject) {
      try {
        this.docker.listImages(function(err, data) {
          log.debug('\n\n:: docker-remote - listImages ::');
          log.debug(data);
          resolve(data);
        });
      } catch (err) {
        log.error(err);
        reject(err);
      }
    }.bind(this));
  },
  getImage: function(imageId) {
    return new Q.Promise(function(resolve, reject) {
      try {
        var image = this.docker.getImage(imageId);
        log.debug('\n\n:: docker-remote - image ::');
        log.debug(image);
        return resolve(image);
      } catch (err) {
        log.error(err);
        reject(err);
      }
    }.bind(this));
  },
  inspectImage: function(image) {
    return new Q.Promise(function(resolve, reject) {
      try {
        var handler = function(err, data) {
          if (err) {
            if (err.statusCode === 404) {
              return resolve(null);
            }
            return reject(err);
          }
          log.debug('\n\n:: docker-remote - image inspect ::');
          log.debug(data);
          return resolve(data);
        };
        image.inspect(handler);
      } catch (err) {
        reject(err);
      }
    }.bind(this));
  },
  removeImage: function(imageId) {
    return Q.async($traceurRuntime.initGeneratorFunction(function $__1() {
      var image,
          handler;
      return $traceurRuntime.createGeneratorInstance(function($ctx) {
        while (true)
          switch ($ctx.state) {
            case 0:
              $ctx.state = 2;
              return this.getImage(imageId);
            case 2:
              image = $ctx.sent;
              $ctx.state = 4;
              break;
            case 4:
              handler = function(err, data) {
                if (err) {
                  throw err;
                }
                log.debug('\n\n:: docker-remote - removeImage ::');
                log.debug(data);
                return data;
              };
              image.remove(handler);
              $ctx.state = -2;
              break;
            default:
              return $ctx.end();
          }
      }, $__1, this);
    }).bind(this))();
  },
  searchImages: function(imageName) {
    return new Q.Promise(function(resolve, reject) {
      try {
        var handler = function(err, data) {
          if (err) {
            return reject(err);
          }
          log.debug('\n\n:: docker-remote - searchImages ::');
          log.debug(data);
          return resolve(data);
        };
        this.docker.searchImages({term: imageName}, handler);
      } catch (err) {
        reject(err);
      }
    }.bind(this));
  },
  searchImagesByTag: function(tagName) {
    return Q.async($traceurRuntime.initGeneratorFunction(function $__1() {
      var allImages,
          imagesFound;
      return $traceurRuntime.createGeneratorInstance(function($ctx) {
        while (true)
          switch ($ctx.state) {
            case 0:
              $ctx.state = 2;
              return this.listImages();
            case 2:
              allImages = $ctx.sent;
              $ctx.state = 4;
              break;
            case 4:
              imagesFound = _.filter(allImages, function(image) {
                if (_.contains(image.RepoTags, tagName)) {
                  return image;
                }
              }, this);
              log.debug('\n\n:: docker-remote - searchImagesByTag ::');
              log.debug(imagesFound);
              $ctx.state = 8;
              break;
            case 8:
              $ctx.returnValue = imagesFound;
              $ctx.state = -2;
              break;
            default:
              return $ctx.end();
          }
      }, $__1, this);
    }).bind(this))();
  },
  getParent: function(imageId) {
    return Q.async($traceurRuntime.initGeneratorFunction(function $__1() {
      var currentImage,
          imageInspect;
      return $traceurRuntime.createGeneratorInstance(function($ctx) {
        while (true)
          switch ($ctx.state) {
            case 0:
              $ctx.state = 2;
              return this.getImage(imageId);
            case 2:
              currentImage = $ctx.sent;
              $ctx.state = 4;
              break;
            case 4:
              $ctx.state = 6;
              return this.inspectImage(currentImage);
            case 6:
              imageInspect = $ctx.sent;
              $ctx.state = 8;
              break;
            case 8:
              $ctx.state = (imageInspect === null) ? 11 : 10;
              break;
            case 11:
              log.debug('\n\n:: docker-remote - getParent ::');
              log.debug('image not found: ' + imageId);
              $ctx.state = 12;
              break;
            case 12:
              $ctx.returnValue = null;
              $ctx.state = -2;
              break;
            case 10:
              log.debug('\n\n:: docker-remote - getParent ::');
              log.debug('image  ID: ' + imageId);
              log.debug('parent ID:   ' + imageInspect.Parent);
              $ctx.state = 17;
              break;
            case 17:
              $ctx.returnValue = ({
                image: currentImage,
                imageInspect: imageInspect
              });
              $ctx.state = -2;
              break;
            default:
              return $ctx.end();
          }
      }, $__1, this);
    }).bind(this))();
  },
  anscestors: function(firstImageId) {
    return Q.async($traceurRuntime.initGeneratorFunction(function $__1() {
      var currentImageId,
          anscestors,
          imageResult,
          imageInspectors;
      return $traceurRuntime.createGeneratorInstance(function($ctx) {
        while (true)
          switch ($ctx.state) {
            case 0:
              currentImageId = firstImageId;
              anscestors = [];
              $ctx.state = 16;
              break;
            case 16:
              $ctx.state = (currentImageId) ? 1 : 12;
              break;
            case 1:
              $ctx.state = 2;
              return this.getParent(currentImageId);
            case 2:
              imageResult = $ctx.sent;
              $ctx.state = 4;
              break;
            case 4:
              $ctx.state = (imageResult === null) ? 7 : 6;
              break;
            case 7:
              log.debug('\n\n:: docker-remote - anscestors ::');
              log.debug('image not found: ' + currentImageId);
              $ctx.state = 8;
              break;
            case 8:
              $ctx.returnValue = null;
              $ctx.state = -2;
              break;
            case 6:
              anscestors.push(imageResult);
              currentImageId = imageResult.imageInspect.Parent;
              $ctx.state = 16;
              break;
            case 12:
              log.debug('\n\n:: docker-remote - anscestors ::');
              imageInspectors = _.pluck(anscestors, 'imageInspect');
              log.debug(_.pluck(imageInspectors, 'Id'));
              $ctx.state = 18;
              break;
            case 18:
              $ctx.returnValue = anscestors;
              $ctx.state = -2;
              break;
            default:
              return $ctx.end();
          }
      }, $__1, this);
    }).bind(this))();
  },
  loadImage: function(outputPath, imageId) {
    return new Q.Promise(function(resolve, reject) {
      try {
        var outputLoadPath = path.join(outputPath, imageId + '.tar');
        var handler = function(err) {
          if (err) {
            return reject(err);
          }
          log.debug('\n\n:: docker-remote - loadImage ::');
          log.debug(outputLoadPath);
          return resolve(outputLoadPath);
        };
        this.docker.loadImage(outputLoadPath, handler);
      } catch (err) {
        reject(err);
      }
    }.bind(this));
  },
  setImageTag: function(namespace, repository, imageId, tagName) {
    return Q.async($traceurRuntime.initGeneratorFunction(function $__1() {
      var image,
          imageInspectResult;
      return $traceurRuntime.createGeneratorInstance(function($ctx) {
        while (true)
          switch ($ctx.state) {
            case 0:
              $ctx.state = 2;
              return this.docker.getImage(imageId);
            case 2:
              image = $ctx.sent;
              $ctx.state = 4;
              break;
            case 4:
              $ctx.state = 6;
              return this.inspectImage(image);
            case 6:
              imageInspectResult = $ctx.sent;
              $ctx.state = 8;
              break;
            case 8:
              $ctx.state = (imageInspectResult === null) ? 11 : 10;
              break;
            case 11:
              log.debug('\n\n:: docker-remote - setImageTag ::');
              log.debug('no image found with ' + imageId);
              $ctx.state = 12;
              break;
            case 12:
              $ctx.returnValue = null;
              $ctx.state = -2;
              break;
            case 10:
              image.tag({
                repo: namespace + '/' + repository,
                tag: tagName,
                force: true
              }, function(err, data) {
                if (err) {
                  throw err;
                }
                return data;
              });
              $ctx.state = -2;
              break;
            default:
              return $ctx.end();
          }
      }, $__1, this);
    }).bind(this))();
  }
}, {});
module.exports = {
  __esModule: true,
  get default() {
    return DockerRemote;
  }
};
