// i18n-cli/lib/index.js
'use strict';

var _bind = Function.prototype.bind;

var _toConsumableArray = function (arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) arr2[i] = arr[i]; return arr2; } else { return Array.from(arr); } };

var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } };

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var path = require('path');
var printf = require('printf');
var chalk = require('chalk');

function load(folder, locale) {
  // jshint ignore:line
  var file = path.join(folder, locale);
  return require(file);
}

var I18n = (function () {
  function I18n(opts) {
    var _this = this;

    _classCallCheck(this, I18n);

    if (typeof opts.dict == 'object') {
      this.dict = opts.dict;
    } else if (opts.locale) {
      this.dict = load(opts.path, opts.locale);
    }

    // Cache gets
    this.cache = {};

    // Alias to translate
    this.t = function () {
      for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      return _this.translate.apply(_this, args);
    };
  }

  _createClass(I18n, [{
    key: '_find',
    value: function _find(keys) {
      var buffer = this.dict || {};

      for (var i = 0; i < keys.length; i++) {
        buffer = buffer[keys[i]];
        if (!buffer) {
          break;
        }
      }

      return buffer;
    }
  }, {
    key: 'translate',
    value: function translate(key) {
      for (var _len2 = arguments.length, args = Array(_len2 > 1 ? _len2 - 1 : 0), _key2 = 1; _key2 < _len2; _key2++) {
        args[_key2 - 1] = arguments[_key2];
      }

      var result = this.cache[key];

      if (typeof result === 'undefined') {
        this.cache[key] = result = this._resolveKey(key);
      }

      if (result.value) {
        try {
          switch (typeof result.value) {
            case 'string':
              return printf.apply(undefined, [result.value].concat(args));
            case 'object':
              return result.value;
            default:
              return key;
          }
        } catch (err) {
          var match,
              label = chalk.red('Translate error');
          match = err.toString().match(/Error: missing key (.*)/);
          if (match) {
            return label + (': \'' + key + '\', missing: ' + match[1]);
          }

          match = err.toString().match(/Error: format requires a mapping/);
          if (match) {
            return label + (': \'' + key + '\', missing a mappping');
          }

          throw err;
        }
      } else {
        return result.key;
      }
    }
  }, {
    key: '_resolveKey',
    value: function _resolveKey(key) {
      var keys = typeof key == 'string' ? key.split('.') : key;
      var value = this._find(keys);

      // Search again, now ancestors is *
      if (!value) {
        var again_keys = new (_bind.apply(Array, [null].concat(_toConsumableArray(keys))))();
        again_keys[again_keys.length - 2] = '*';
        value = this._find(again_keys);
      }

      // Key to show in a error
      key = chalk.yellow(typeof key == 'string' ? key : key.join('.'));

      return { key: key, value: value };
    }
  }]);

  return I18n;
})();

// Support es6 and es5
I18n.I18n = I18n;
I18n['default'] = I18n;
I18n.__esModule = true;
module.exports = I18n;
