// docker-registry-downloader/lib/src/fs-helper/index.js
"use strict";
var Q = require('q');
var fs = require('fs');
var FS = require("q-io/fs");
var rmdir = require('rimraf');
var FsHelper = function FsHelper() {};
($traceurRuntime.createClass)(FsHelper, {
  removeDirRecursive: function(dir) {
    return new Q.Promise(function(resolve, reject) {
      rmdir(dir, function(err) {
        if (err) {
          reject(err);
        } else {
          resolve(true);
        }
      });
    });
  },
  createCleanFolder: function(fullPath) {
    return Q.async($traceurRuntime.initGeneratorFunction(function $__1() {
      var $__2,
          $__3,
          $__4;
      return $traceurRuntime.createGeneratorInstance(function($ctx) {
        while (true)
          switch ($ctx.state) {
            case 0:
              $__2 = this.fsExists;
              $__3 = $__2.call(this, fullPath);
              $ctx.state = 6;
              break;
            case 6:
              $ctx.state = 2;
              return $__3;
            case 2:
              $__4 = $ctx.sent;
              $ctx.state = 4;
              break;
            case 4:
              $ctx.state = ($__4) ? 7 : 10;
              break;
            case 7:
              $ctx.state = 8;
              return this.removeDirRecursive(fullPath);
            case 8:
              $ctx.maybeThrow();
              $ctx.state = 10;
              break;
            case 10:
              $ctx.state = 13;
              return FS.makeTree(fullPath);
            case 13:
              $ctx.maybeThrow();
              $ctx.state = 15;
              break;
            case 15:
              $ctx.returnValue = fullPath;
              $ctx.state = -2;
              break;
            default:
              return $ctx.end();
          }
      }, $__1, this);
    }).bind(this))();
  },
  createFolder: function(fullPath) {
    return Q.async($traceurRuntime.initGeneratorFunction(function $__1() {
      return $traceurRuntime.createGeneratorInstance(function($ctx) {
        while (true)
          switch ($ctx.state) {
            case 0:
              $ctx.state = 2;
              return FS.makeTree(fullPath);
            case 2:
              $ctx.maybeThrow();
              $ctx.state = 4;
              break;
            case 4:
              $ctx.returnValue = fullPath;
              $ctx.state = -2;
              break;
            default:
              return $ctx.end();
          }
      }, $__1, this);
    }).bind(this))();
  },
  fsExists: function(fullPath) {
    return new Q.Promise(function(resolve, reject) {
      try {
        fs.exists(fullPath, function(result) {
          resolve(result);
        });
      } catch (err) {
        reject(err);
      }
    }.bind(this));
  },
  tarPack: function(folderToPack, outputTarfile) {
    return new Q.Promise(function(resolve, reject) {
      try {
        var write = fs.createWriteStream;
        var pack = require('tar-pack').pack;
        pack(folderToPack).pipe(write(outputTarfile)).on('error', function(err) {
          reject(err);
        }).on('close', function() {
          resolve(true);
        });
      } catch (err) {
        reject(err);
      }
    }.bind(this));
  }
}, {});
module.exports = {
  __esModule: true,
  get default() {
    return FsHelper;
  }
};
